fdbserver/BlobManager.actor.cpp

```diff
@@ -873,38 +840,79 @@ ACTOR Future<Void> blobManager(LocalityData locality, Reference<AsyncVar<ServerD
 		printf("Blob manager taking lock\n");
 	}
 
-	int64_t _epoch = wait(acquireManagerLock(&self));
-	self.epoch = _epoch;
+	self.epoch = epoch;
+
+	// make sure the epoch hasn't gotten stale
+	state Reference<ReadYourWritesTransaction> tr = makeReference<ReadYourWritesTransaction>(self.db);
+
+	loop {
+		tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+		tr->setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+		try {
+			wait(checkManagerLock(tr, &self));
+			wait(tr->commit());
+			break;
+		} catch (Error& e) {
+			if (e.code() == error_code_granule_assignment_conflict) {
+				if (BM_DEBUG) {
+					printf("Blob manager dying...\n");
+				}
+				return Void();
+			}
+
+			// if we get here, most likely error is a read-write conflict
+			if (BM_DEBUG) {
+				printf("Blob manager lock check got unexpected error %s\n", e.name());
+			}
+			wait(tr->onError(e));
+		}
+	}
+
 	if (BM_DEBUG) {
-		printf("Blob manager acquired lock at epoch %lld\n", _epoch);
+		printf("Blob manager acquired lock at epoch %lld\n", epoch);
 	}
 
 	int numWorkers = 2;
 	for (int i = 0; i < numWorkers; i++) {
-		state BlobWorkerInterface bwInterf(locality, deterministicRandom()->randomUniqueID());
+		state BlobWorkerInterface bwInterf(bmInterf.locality, deterministicRandom()->randomUniqueID());
 		bwInterf.initEndpoints();
 		self.workersById.insert({ bwInterf.id(), bwInterf });
 		self.workerStats.insert({ bwInterf.id(), BlobWorkerStats() });
 		self.workerMonitors.insert({ bwInterf.id(), monitorBlobWorker(&self, bwInterf) });
-		addActor.send(blobWorker(bwInterf, dbInfo));
+		self.addActor.send(blobWorker(bwInterf, dbInfo));
 	}
 
-	addActor.send(monitorClientRanges(&self));
-	addActor.send(rangeAssigner(&self));
+	self.addActor.send(monitorClientRanges(&self));
+	self.addActor.send(rangeAssigner(&self));
 
 	if (BUGGIFY) {
-		addActor.send(rangeMover(&self));
+		self.addActor.send(rangeMover(&self));
 	}
 
 	// TODO probably other things here eventually
-	loop choose {
-		when(wait(self.iAmReplaced.getFuture())) {
-			if (BM_DEBUG) {
-				printf("Blob Manager exiting because it is replaced\n");
+	try {
+		loop choose {
+			when(wait(self.iAmReplaced.getFuture())) {
+				if (BM_DEBUG) {
+					printf("Blob Manager exiting because it is replaced\n");
+				}
+				return Void();
+			}
+			when(HaltBlobManagerRequest req = waitNext(bmInterf.haltBlobManager.getFuture())) {
+				req.reply.send(Void());
+				TraceEvent("BlobManagerHalted", bmInterf.id()).detail("ReqID", req.requesterID);
+				break;
```

Comment:
This is a minor nit, but iAmReplaced returns directly and the halt request breaks and then returns. We should probably be consistent with which one we do in both places

Reply:
Thanks for noticing that!